### PR TITLE
Implemented offset information and major refactoring of tokenisation functions.

### PR DESCRIPTION
**Pull request by proycon**
This (draft) PR is the result of #14 , it implements offset support for **all** the tokenisers.

As a lot of the original logic consisted of simple string replacements, losing the alignment with the original text, various internal functions had to be modified or rewritten completely. This code therefore constitutes a major refactoring. All logic is rewritten to use new ``Token`` (or ``TokenRef``, holding ``&str`` instead of ``String``) structures. Each token holds an offset and a mask. The mask tells something about the role of the token (is it a special token? out of vocabulary? is it a continuation subtoken?).

The core of the tokenisers is formed by the following new shared functions in ``tokenization_utils``:

* ``split_on_char()`` - split on character (takes a testing function as argument)
* ``split_on_substr()`` - split on substring (takes a testing function as argument)
* ``split_on_regex()`` - split on regular expression (takes a regex as argument)
* ``split_on_bpe_pairs()`` - split on bpe pairs, this abstracts over parts of the GPT2/Roberta/ctrl tokenizers (takes a bpe function as argument).

The various higher-order functions all use these:

* ``split_on_regex_with_lookahead()`` - abstracts a part of the GPT2/Roberta tokenizers
* ``split_on_punct``- split on punctuation
* ``whitespace_tokenize``- split on whitespace
* ``tokenize_cjk_chars()``- split on CJK characters

All these function take a single Token or TokenRef (which may correspond to an entire text/sentence/whatever!) and return either ``Vec<Token>`` or ``Vec<TokenRef>`` (depending on whether I could get away with not doing any copies). The general idea is that tokens are being constantly decomposed into smaller subtokens by chaining the functions in an appropriate order.

Each of the tokenisers has a ``tokenize_into_tokens() -> Vec<Token>`` method coming from the Tokenizer trait. ``tokenize_with_offsets()`` is a small wrapper around that which decomposes the ``Vec<Token>`` into a 3-tuple ``(Vec<String>,Vec<Offset>,Vec<Mask>)``. The original ``tokenize()`` function remains for backward compatibility but simply ignores the offset and mask information (so has no performance benefits).

As for performance, though certain things got optimized and will surely help, there is also added complexity that will have some impact. Things can probably still be improved (and cleaned up) too.

Because of all the extensive changes, the deeper code no longer maps one-on-one to the original transformer's tokenizers (assuming it did in the first place?), still the behaviour should be almost identical.

This is a draft pull request at this stage, because there are still some loose ends to fix. The main issue is that I currently have failures on a few tests which I'm not sure about (I'll address it in a new comment because this one is already getting pretty long).

### CODE REVIEW

**Comment by proycon** [comment]
@guillaume-be The last issue I'm running into concerns some failing tests (all related)  where I'm not sure if the original test reference output is best. The failures all relate to the BPE-like tokenizers when they get a only whitespace as input, the test reference expects a single UNK token as output. My implementation simply returns no tokens at all (empty vector), which  seems pretty sensible to me as there are no tokens to work on. The question is if you really require the old behaviour in this corner-case?

failures:

```
---- preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer_no_lower_casing stdout ----
thread 'preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer_no_lower_casing' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `["<unk>"]`', src/preprocessing/tokenizer/ctrl_tokenizer.rs:202:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer stdout ----
thread 'preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `["<unk>"]`', src/preprocessing/tokenizer/ctrl_tokenizer.rs:163:13

---- preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_encode stdout ----
thread 'preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_encode' panicked at 'assertion failed: `(left == right)`
  left: `TokenizedInput { token_ids: [], segment_ids: [], special_tokens_mask: [], overflowing_tokens: [], num_truncated_tokens: 0, token_offsets: [], mask: [] }`,
 right: `TokenizedInput { token_ids: [6], segment_ids: [0], special_tokens_mask: [0], overflowing_tokens: [], num_truncated_tokens: 0, token_offsets: [], mask: [] }`', src/preprocessing/tokenizer/openai_gpt_tokenizer.rs:228:13

---- preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer stdout ----
thread 'preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `["<unk>"]`', src/preprocessing/tokenizer/openai_gpt_tokenizer.rs:151:13

---- preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer_no_lower_casing stdout ----
thread 'preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer_no_lower_casing' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `["<unk>"]`', src/preprocessing/tokenizer/openai_gpt_tokenizer.rs:186:13

---- preprocessing::tokenizer::gpt2_tokenizer::tests::test_encode stdout ----
thread 'preprocessing::tokenizer::gpt2_tokenizer::tests::test_encode' panicked at 'assertion failed: `(left == right)`
  left: `TokenizedInput { token_ids: [5], segment_ids: [0], special_tokens_mask: [0], overflowing_tokens: [], num_truncated_tokens: 0, token_offsets: [Some(Offset { begin: 0, end: 1 })], mask: [Begin] }`,
 right: `TokenizedInput { token_ids: [6], segment_ids: [0], special_tokens_mask: [0], overflowing_tokens: [], num_truncated_tokens: 0, token_offsets: [Some(Offset { begin: 0, end: 1 })], mask: [None] }`', src/preprocessing/tokenizer/gpt2_tokenizer.rs:248:1
3

---- preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer stdout ----
thread 'preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer' panicked at 'assertion failed: `(left == right)`
  left: `["Ġ"]`,
 right: `["<|endoftext|>"]`', src/preprocessing/tokenizer/gpt2_tokenizer.rs:177:13

---- preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer_no_lower_casing stdout ----
thread 'preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer_no_lower_casing' panicked at 'assertion failed: `(left == right)`
  left: `["Ġ"]`,
 right: `["<|endoftext|>"]`', src/preprocessing/tokenizer/gpt2_tokenizer.rs:212:13


failures:
    preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer
    preprocessing::tokenizer::ctrl_tokenizer::tests::test_ctrl_tokenizer_no_lower_casing
    preprocessing::tokenizer::gpt2_tokenizer::tests::test_encode
    preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer
    preprocessing::tokenizer::gpt2_tokenizer::tests::test_gpt2_tokenizer_no_lower_casing
    preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_encode
    preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer
    preprocessing::tokenizer::openai_gpt_tokenizer::tests::test_openai_gpt_tokenizer_no_lower_casing
```

**Comment by guillaume-be** [comment]
Hello,

Thank you this is amazing work! I am reviewing the changes (will take some time given the amount of changes). The concept of Token and TokenRef are great, they make the code a lot cleaner. You are right that these changes deviate from the original implementation, but the code becomes more Rust-idiomatic which is a good trade-off. Staying close to the original implementation was helpful as a first port for ensuring the output is matching was is expected.

I am making some changes directly on the code, I will open a PR on your branch to integrate these today.

I have already looked into the following:
- unit test fixes
- benchmarking
- splitting of special token for BPE-based tokenizers
- minor code cleanup (eliding lifetimes, typos)

I am also running the integration tests and there are still a few edge cases that are failing, I am working on these.

Thanks again!

**Comment by guillaume-be** [comment]
@proycon I have all unit tests and integration tests fixed. Eventually only minor changes were required to have everything work out. I am unsure how I could push these changes to the current PR, could you please help me out?

**Comment by guillaume-be** [comment]
@proycon Thanks again for the PR. I have pushed some changes directly to this PR. Please feel free to review and let me know what you think.

All unit tests and integration tests are now running. I guess you could publish this PR.
I have run the benchmark and compared to the old version, and the performance is identical. So you managed to add significant features at no additional overhead, congratulations!

I would still like to take some more time later today to review, but all in all this looks fantastic!

**Comment by proycon** [comment]
Thanks!! I see you managed to push them to the current PR, I'll take a look!

**Comment by proycon** [comment]
I'm also glad to hear the performance is nearly identical! I had done no benchmarks whatsoever myself.

**Comment by proycon** [comment]
The changes look good, thanks for cleaning up some of the formatting, I hadn't gotten around to that yet, especially in the tests things could be a bit messy. Good to see all the tests are green now! I marked the PR ready for review/merge.

File: main/src/preprocessing/tokenizer/base_tokenizer.rs [hunk 1]
```diff
@@ -328,7 +324,7 @@ pub trait Tokenizer<T: Vocab> {
             .replace(" ,", ",")
             .replace(" ' ", "'")
             .replace(" n't", "n't")
-            .replace(" 'm", "'m")
+            .replace(" 'm'", "'m")
```
Comment (proycon) [inline main/src/preprocessing/tokenizer/base_tokenizer.rs:327]: Ah? So it wasn't a typo after all? My rust-be tests are green now with this re-commit!

Comment (guillaume-be) [reply main/src/preprocessing/tokenizer/base_tokenizer.rs:327]: It was a typo. The tests for `rust-bert` probably need to be updated. I am about to push some changes that re-re-fixes the typo you found in a previous PR but got lost. Sorry that is what I meant.

**Comment by proycon** [comment]
I tried to make the interface for Token and TokenRef a bit more elegant and idiomatic:
 * conversion between the two can be done using the From trait
   * the explicit conversion function have been renamed to be a bit more idiomatic
 * both structs implement a new TokenTrait  (this will be useful later in rust-bert where the Token we defined there can implement the same trait).

I figured I'd best add it to this PR rather than open a new one, as this was still open, but let me know if you prefer otherwise.
